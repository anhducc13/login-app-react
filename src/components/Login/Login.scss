.Login {
  position: absolute;
  top: 0;
  left: 0;
  height: 100vh;
  width: 100%;
  z-index: 1;
}

.form-box {
  margin-top: 60px;
  background-color: #ffffff;
  z-index: 10;
  padding: 10px 20px;
  border-radius: 10px;
  box-shadow: 0 1px 4px 0 rgba(0, 0, 0, 0.5);
  flex-direction: column;
}

.social-divider {
  width: 90%;
  text-align: center;
  border-bottom: 1px solid #979faf;
  line-height: 0.1em;
  margin: 5px auto 20px;
}

.social-divider span {
  background: #fff;
  padding: 0 10px;
}
