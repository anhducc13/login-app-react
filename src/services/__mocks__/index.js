// @flow
export { default as authServices } from './auth';