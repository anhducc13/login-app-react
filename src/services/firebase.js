import * as firebase from 'firebase';
import uuidv4 from 'uuid/v4';

const appFirebase = firebase.initializeApp({
    apiKey: "AIzaSyAcjbrVQyOOlY-SadQ8TWmldob2SjH5trM",
    authDomain: "ductt-217110.firebaseapp.com",
    databaseURL: "https://ductt-217110.firebaseio.com",
    storageBucket: "ductt-217110.appspot.com",
});

const uploadFileImage = (blob) => {
    const filename = Date.now().toString() + uuidv4();
    const ref = appFirebase.storage().ref(`avatar/${filename}`);
    return ref.put(blob).then(snapshot => snapshot.metadata.fullPath)
}

const getDownloadURL = (refName) => {
    const ref = appFirebase.storage().ref();
    return ref.child(refName).getDownloadURL();
}

const deleteFileImage = (refName) => {
    const ref = appFirebase.storage().ref();
    return ref.child(refName).delete();
}

export default {
    uploadFileImage,
    getDownloadURL,
    deleteFileImage,
};