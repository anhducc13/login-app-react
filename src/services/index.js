// @flow
export { default as userServices } from './user';
export { default as authServices } from './auth';
export { default as requestServices } from './requests';
export { default as firebaseServices } from './firebase';
