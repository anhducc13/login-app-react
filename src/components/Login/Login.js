import React, { useState, useContext } from 'react';
import {
  Form,
  Input,
  Button,
  Col,
  Row,
} from 'antd';
import { Link } from 'react-router-dom';
import './Login.scss';
import Title from 'antd/lib/typography/Title';
import { authServices } from 'services';
import openNotificationWithIcon from 'helpers/notification';
import { UserContext } from 'UserContext';

const { Password } = Input;


const Login = (props) => {

  const bodyLayout = {
    xs: { span: 22, offset: 1 },
    sm: { span: 16, offset: 4 },
    md: { span: 12, offset: 6 },
    lg: { span: 10, offset: 7 },
    xl: { span: 8, offset: 8 },
    xxl: { span: 6, offset: 9 },
  }

  const [loading, setLoading] = useState(false);

  const [username, setUsername] = useState("")
  const [password, setPassword] = useState("")
  const [helperUsername, setHelperUsername] = useState("")
  const [helperPassword, setHelperPassword] = useState("")
  const [errorText, setErrorText] = useState("")

  const [, setUser] = useContext(UserContext);

  const validateInput = () => {
    if (username === "")
      setHelperUsername("Please enter username");
    else
      setHelperUsername("");

    if (password === "")
      setHelperPassword("Please enter password")
    else
      setHelperPassword("");
  }

  const handleSubmit = (e) => {
    e.preventDefault();
    validateInput();
    setErrorText("")
    if (username !== "" && password !== "") {
      setLoading(true);
      const params = {
        "username": username,
        "password": password,
      }
      authServices.loginUser(params)
        .then(res => {
          setLoading(false);
          const { data } = res;
          setUser(data);
          openNotificationWithIcon('success', 'Success', `Welcome back, ${data.username}!`)
          props.history.push('/');
        })
        .catch(err => {
          setLoading(false);
          if (err && err.response) {
            const { status, data } = err.response;
            if (status === 400) {
              setErrorText(data.message)
              return;
            }
            if (status === 404) {
              props.history.push('/404');
              return;
            }
            if (status === 403) {
              if (data.message === "Account warning") {
                setErrorText("You enter wrong password 3 times. 5 times, account will be block")
                return;
              }
              setErrorText("Account has been block. Please try after")
              return;
            }
            if (status > 400 && status < 500) {
              props.history.push('/403');
              return;
            }
          }
          props.history.push('/500');
        })
    }
  }

  return (
    <div className="Login">
      <Col {...bodyLayout}>
        <Form className="form-box">
          <Title
            level={2}
            style={{
              padding: "5px",
              textAlign: "center"
            }}
          >
            Login
          </Title>
          <Title
            level={4}
            type="danger"
            style={{
              textAlign: "center"
            }}
          >
            {errorText}
          </Title>
          <Form.Item label="Username" help={helperUsername} validateStatus={helperUsername && "error"}>
            <Input size="large" name="username" value={username} onChange={(e) => setUsername(e.target.value)} />
          </Form.Item>
          <Form.Item label="Password" help={helperPassword} validateStatus={helperPassword && "error"}>
            <Password size="large" name="password" value={password} onChange={(e) => setPassword(e.target.value)} />
          </Form.Item>
          <Row>
            <Col span={12}>
              <Link to="/forgot-password">Forgot password?</Link>
            </Col>
            <Col span={12} style={{ textAlign: "right" }}>
              <Link to="/register">Don't have an account?</Link>
            </Col>
          </Row>
          <Form.Item>
            <Button
              type="primary"
              htmlType="submit"
              block
              loading={loading}
              onClick={handleSubmit}
              size="large"
              style={{
                marginTop: 10,
              }}
            >
              Login
            </Button>
          </Form.Item>
        </Form>
      </Col>
    </div>
  );
};


export default Login;
