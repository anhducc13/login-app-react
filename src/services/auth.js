import { requestServices } from 'services';

const loginUser = (payload) => requestServices.customAxios.post('/auth/login',payload)

const registerUser = (payload) => requestServices.customAxios.post('/auth/register',payload)

const forgotPasswordUser = (payload) => requestServices.customAxios.post('/auth/forgotPassword',payload)

const updatePasswordUser = (payload) => requestServices.customAxios.post('/auth/updatePassword',payload)

const logoutUser = () => requestServices.customAxios.get('/auth/logout')

const currentUser = () => requestServices.customAxios.get('/auth/currentUser')


  export default {
    loginUser,
    currentUser,
    registerUser,
    forgotPasswordUser,
    logoutUser,
    updatePasswordUser
  };