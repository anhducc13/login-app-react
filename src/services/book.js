// @flow
import { requestServices } from 'services';

const fetchBooks = (args) => requestServices.customAxios.get('/book/list/', { params: args })
  .then(res => res.data.data);

const getAllCategory = () => requestServices.customAxios.get(`/book/allCategory`)
  .then(res => res.data.data);

const fetchCategoryActions = (args) => requestServices.customAxios.get(`/category/activity`, { params: args })
  .then(res => res.data.data);

const addBook = (params) => requestServices.customAxios.post('/book/', params, {headers: {
  'Content-Type': 'application/json',
}})
  .then(res => res.data.data);

const editCategory = (id, params) => requestServices.customAxios.put(`/category/${id}`, params)
  .then(res => res.data.data);

const deleteCategory = (id) => requestServices.customAxios.delete(`/category/${id}`)
  .then(res => res.data.data);

export default {
  fetchBooks,
  getAllCategory,
  addBook,
  editCategory,
  deleteCategory,
  fetchCategoryActions,
};
