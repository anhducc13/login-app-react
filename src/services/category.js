// @flow
import { requestServices } from 'services';

const fetchCategories = (args) => requestServices.customAxios.get('/category/list/', { params: args })
  .then(res => res.data.data);

const fetchCategory = (id) => requestServices.customAxios.get(`/category/${id}`)
  .then(res => res.data.data);

const addCategory = (params) => requestServices.customAxios.post('/category/', params)
  .then(res => res.data.data);

const editCategory = (id, params) => requestServices.customAxios.put(`/category/${id}`, params)
  .then(res => res.data.data);

const deleteCategory = (id) => requestServices.customAxios.delete(`/category/${id}`)
  .then(res => res.data.data);

export default {
  fetchCategories,
  fetchCategory,
  addCategory,
  editCategory,
  deleteCategory,
};
