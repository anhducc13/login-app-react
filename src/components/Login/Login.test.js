import React from 'react';
import { mount, shallow } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { authServices } from 'services';
import Login from './Login';

describe('Login Page Render', () => {
  it('should render correctly', () => {
    const component = shallow(<Login />);
    expect(component).toMatchSnapshot();
  });
});

describe('Login form validate', () => {
  it('validate test case 1', () => {
    const component = mount(
      <MemoryRouter>
        <Login />
      </MemoryRouter>
    );
    component.find('input[name="username"]').simulate('change', { target: { value: "" } });
    component.find('button[type="submit"]').simulate('click');
    expect(component.find('div.ant-form-explain').first().text())
      .toBe("Please enter username")
    component.unmount()
  })

  it('validate test case 2', () => {
    const component = mount(
      <MemoryRouter>
        <Login />
      </MemoryRouter>
    );
    component.find('input[name="username"]').simulate('change', { target: { value: "anhducc14" } });
    component.find('input[name="password"]').simulate('change', { target: { value: "" } });
    component.find('button[type="submit"]').simulate('click');
    expect(component.find('div.ant-form-explain').length)
      .toBe(1)
    expect(component.find('div.ant-form-explain').first().text())
      .toBe("Please enter password")

    component.unmount()
  })

  it('validate test case 3', () => {
    const component = mount(
      <MemoryRouter>
        <Login />
      </MemoryRouter>
    );
    component.find('input[name="username"]').simulate('change', { target: { value: "anhducc14" } });
    component.find('input[name="password"]').simulate('change', { target: { value: "Anhducc14" } });
    component.find('button[type="submit"]').simulate('click');
    expect(component.find('div.ant-form-explain').length)
      .toBe(0)
    component.unmount()
  })
});


describe('Login Page Submit Form', () => {
  let loginSpy;

  beforeAll(() => {
    loginSpy = jest.fn();
    loginSpy = jest.spyOn(authServices, 'loginUser');
  });

  it('Submit login will not call loginUser function after fail validate username', () => {
    const component = mount(
      <MemoryRouter>
        <Login />
      </MemoryRouter>
    );
    component.find('input[name="username"]').simulate('change', { target: { value: "" } });
    component.find('input[name="password"]').simulate('change', { target: { value: "Anhducc14" } });
    component.find('button[type="submit"]').simulate('click');
    expect(loginSpy).not.toHaveBeenCalled()
    component.unmount()
  })

  it('Submit login will not call loginUser function after fail validate password', () => {
    const component = mount(
      <MemoryRouter>
        <Login />
      </MemoryRouter>
    );
    component.find('input[name="username"]').simulate('change', { target: { value: "a" } });
    component.find('input[name="password"]').simulate('change', { target: { value: "" } });
    component.find('button[type="submit"]').simulate('click');
    expect(loginSpy).not.toHaveBeenCalled()
    component.unmount()
  })

  it('Submit login will call loginUser function after success validate',() => {
    const component = mount(
      <MemoryRouter>
        <Login />
      </MemoryRouter>
    );
    component.find('input[name="username"]').simulate('change', { target: { value: "a" } });
    component.find('input[name="password"]').simulate('change', { target: { value: "1" } });
    component.find('button[type="submit"]').simulate('click');
    expect(loginSpy)
      .toHaveBeenCalled()
    expect(loginSpy)
      .toHaveBeenCalledWith({
        "username": "a",
        "password": "1"
      })
    component.unmount()
  })

  afterAll(() => {
    jest.restoreAllMocks();
  });
})