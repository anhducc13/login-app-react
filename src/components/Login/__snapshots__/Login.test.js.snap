// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page Render should render correctly 1`] = `
<div
  className="Login"
>
  <Col
    lg={
      Object {
        "offset": 7,
        "span": 10,
      }
    }
    md={
      Object {
        "offset": 6,
        "span": 12,
      }
    }
    sm={
      Object {
        "offset": 4,
        "span": 16,
      }
    }
    xl={
      Object {
        "offset": 8,
        "span": 8,
      }
    }
    xs={
      Object {
        "offset": 1,
        "span": 22,
      }
    }
    xxl={
      Object {
        "offset": 9,
        "span": 6,
      }
    }
  >
    <Form
      className="form-box"
      colon={true}
      hideRequiredMark={false}
      layout="horizontal"
      onSubmit={[Function]}
    >
      <Title
        level={2}
        style={
          Object {
            "padding": "5px",
            "textAlign": "center",
          }
        }
      >
        Login
      </Title>
      <FormItem
        hasFeedback={false}
        help=""
        label="Username"
        validateStatus=""
      >
        <Input
          disabled={false}
          name="username"
          onChange={[Function]}
          size="large"
          type="text"
          value=""
        />
      </FormItem>
      <FormItem
        hasFeedback={false}
        help=""
        label="Password"
        validateStatus=""
      >
        <Password
          action="click"
          inputPrefixCls="ant-input"
          name="password"
          onChange={[Function]}
          prefixCls="ant-input-password"
          size="large"
          value=""
          visibilityToggle={true}
        />
      </FormItem>
      <Row
        gutter={0}
      >
        <Col
          span={12}
        >
          <Link
            to="/forgot-password"
          >
            Forgot password?
          </Link>
        </Col>
        <Col
          span={12}
          style={
            Object {
              "textAlign": "right",
            }
          }
        >
          <Link
            to="/register"
          >
            Don't have an account?
          </Link>
        </Col>
      </Row>
      <FormItem
        hasFeedback={false}
      >
        <Button
          block={true}
          ghost={false}
          htmlType="submit"
          loading={false}
          onClick={[Function]}
          size="large"
          style={
            Object {
              "marginTop": 10,
            }
          }
          type="primary"
        >
          Login
        </Button>
      </FormItem>
      <FormItem
        hasFeedback={false}
      >
        <div
          className="social-divider"
        >
          <span>
            or
          </span>
        </div>
        <e
          accessType="online"
          buttonText="Sign in with Google"
          clientId="335058615265-8prgp3oprps9sucnlubbs7rc3slgat1m.apps.googleusercontent.com"
          cookiePolicy="single_host_origin"
          disabledStyle={
            Object {
              "opacity": 0.6,
            }
          }
          fetchBasicProfile={true}
          icon={true}
          isSignedIn={false}
          jsSrc="https://apis.google.com/js/api.js"
          onFailure={[Function]}
          onRequest={[Function]}
          onSuccess={[Function]}
          prompt=""
          render={[Function]}
          scope="profile email"
          tag="button"
          theme="light"
          type="button"
          uxMode="popup"
        />
      </FormItem>
    </Form>
  </Col>
</div>
`;
